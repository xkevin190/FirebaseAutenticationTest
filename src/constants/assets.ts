export const FONTS = {
    LATO: 'Lato',
};