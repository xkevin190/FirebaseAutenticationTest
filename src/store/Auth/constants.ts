export type IAccountActionTypes = {
  CREATE_ACCOUNT: string;
  LOGIN: string;
  ClOSE_SESSION: string;
};

export const AccounthActionsTypes: IAccountActionTypes = {
  CREATE_ACCOUNT: 'CREATE_ACCOUNT',
  LOGIN: ' LOGIN',
  ClOSE_SESSION: 'ClOSE_SESSION',
};
