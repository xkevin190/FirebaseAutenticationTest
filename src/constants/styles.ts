import { FONTS } from "./assets";

export const COLORS = {
    WHITE: '#FFFFFF',
    BLACK: '#000000',
    BLACK_025: '#00000040',
    GRAY_DEFAULT: '#46484B',
    GRAY_DEFAULT_DARKER: '#646565',
    GRAY_MEDIUM: '#9DA5B9',
    GRAY_LIGHT: '#F6F8FF',
    GRAY_LIGHT_DARKER: '#DDDDDD',
    DISABLED_DEFAULT: '#DFE6ED',
    OPYA_BLUE: '#2162E3',
    OPYA_VIOLET: '#1745D5',
    MAIN_DEFAULT: '#313342',
    SECONDARY_LIGHT: '#62698B',
    ALERT_DEFAULT: '#D15686',
    ALERT_LIGHT: '#FFD0E2',
  };

  const FONTS_TYPES = {
    BLACK: 'Black',
    BLACK_ITALIC: 'BlackItalic',
    BOLD: 'Bold',
    BOLD_ITALIC: 'BoldItalic',
    HAIR_LINE: 'HairLine',
    HAIR_LINE_ITALIC: 'HairLineItalic',
    HEAVY: 'Heavy',
    HEAVY_ITALIC: 'HeavyItalic',
    ITALIC: 'Italic',
    LIGHT: 'Light',
    LIGHT_ITALIC: 'LightItalic',
    MEDIUM: 'Medium',
    MEDIUM_ITALIC: 'MediumItalic',
    REGULAR: 'Regular',
    SEMI_BOLD: 'SemiBold',
    SEMI_BOLD_ITALIC: 'SemiBoldItalic',
    THIN: 'Thin',
    THIN_ITALIC: 'ThinItalic',
  };


  export const STYLES = {
    FONT_FAMILY: {
      LATO_REGULAR: `${FONTS.LATO}-${FONTS_TYPES.REGULAR}`,
      LATO_BOLD: `${FONTS.LATO}-${FONTS_TYPES.BOLD}`,
    },
    FONT_SIZE: {
      PARAGRAPH: 14,
      SMALL_PARAGRAPH: 12,
      MEDIUM_PARAGRAPH: 16,
      INTERMEDIUM_PARAGRAPH: 18,
      LARGE_PARAGRAPH: 20,
    },
  };
  

