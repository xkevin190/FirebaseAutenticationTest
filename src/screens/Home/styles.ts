import {StyleSheet, ViewStyle} from 'react-native';

import {COLORS} from '../../constants/styles';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'column',
    backgroundColor: COLORS.GRAY_LIGHT,
  } as ViewStyle,
});

export default styles;
