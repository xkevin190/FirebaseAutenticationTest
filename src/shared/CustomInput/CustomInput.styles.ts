import {StyleSheet, TextStyle, ViewStyle} from 'react-native';

import {COLORS, STYLES} from '../../constants/styles';

const styles = StyleSheet.create({
  inputContainer: {
    width: '100%',
    backgroundColor: COLORS.WHITE,
    borderRadius: 16,
    borderWidth: 1,
    marginTop: 6,
    marginBottom: 18,
    flexDirection: 'row',
  } as ViewStyle,
  input: {
    fontFamily: STYLES.FONT_FAMILY.LATO_REGULAR,
    fontSize: STYLES.FONT_SIZE.PARAGRAPH,
    paddingVertical: 10,
    paddingHorizontal: 16,
    flex: 1,
  } as TextStyle,
  iconContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    width: '15%',
  },
  errorContainer: {
    marginBottom: 18,
  } as ViewStyle,
});

export default styles;
