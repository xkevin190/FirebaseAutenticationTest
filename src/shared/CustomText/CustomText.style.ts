import {StyleSheet, TextStyle} from 'react-native';

import {STYLES} from '../../constants/styles';

const styles = StyleSheet.create({
  'small-paragraph-bold': {
    fontFamily: STYLES.FONT_FAMILY.LATO_BOLD,
    fontSize: STYLES.FONT_SIZE.SMALL_PARAGRAPH,
  } as TextStyle,
  'small-paragraph': {
    fontFamily: STYLES.FONT_FAMILY.LATO_REGULAR,
    fontSize: STYLES.FONT_SIZE.SMALL_PARAGRAPH,
  } as TextStyle,
  'medium-paragraph': {
    fontFamily: STYLES.FONT_FAMILY.LATO_REGULAR,
    fontSize: STYLES.FONT_SIZE.MEDIUM_PARAGRAPH,
  } as TextStyle,
  paragraph: {
    fontFamily: STYLES.FONT_FAMILY.LATO_REGULAR,
    fontSize: STYLES.FONT_SIZE.PARAGRAPH,
  } as TextStyle,
  'large-paragraph': {
    fontFamily: STYLES.FONT_FAMILY.LATO_REGULAR,
    fontSize: STYLES.FONT_SIZE.LARGE_PARAGRAPH,
  } as TextStyle,
  'paragraph-bold': {
    fontFamily: STYLES.FONT_FAMILY.LATO_BOLD,
    fontSize: STYLES.FONT_SIZE.PARAGRAPH,
  } as TextStyle,
  'intermedium-paragraph-bold': {
    fontFamily: STYLES.FONT_FAMILY.LATO_BOLD,
    fontSize: STYLES.FONT_SIZE.INTERMEDIUM_PARAGRAPH,
  } as TextStyle,
  'intermedium-paragraph': {
    fontSize: STYLES.FONT_SIZE.INTERMEDIUM_PARAGRAPH,
    fontFamily: STYLES.FONT_FAMILY.LATO_REGULAR,
  } as TextStyle,
});

export default styles;
