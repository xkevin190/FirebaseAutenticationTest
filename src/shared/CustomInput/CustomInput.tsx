import React, { useState } from 'react';
import { TextInput, View } from 'react-native';
import { useTranslation } from 'react-i18next';

import CustomText from '../CustomText';
import { COLORS } from '../../constants/styles';
import styles from './CustomInput.styles';

export interface ICustomInputProps {
  title?: string;
  validationsText?: string[];
  placeholder: string;
  value: string;
  height?: number;
  textArea?: boolean;
  pattern?: RegExp[];
  required?: boolean;
  password?: boolean;
  icon?: React.ReactNode;
  onChangeText: (text: string) => void;
}

const CustomInput: React.FC<ICustomInputProps> = ({
  title = '',
  placeholder,
  value,
  validationsText = [],
  height,
  textArea = false,
  pattern,
  required = false,
  password = false,
  icon = null,
  onChangeText,
}) => {
  const { t } = useTranslation();
  const [hasError, setHasError] = useState(false);
  const isRequiredEnabled = required && !value;
  const isValidationEnabled = hasError && !!value;
  const isErrorEnabled = isRequiredEnabled || isValidationEnabled;
  const handleValidation = (valueToValidate: string): boolean => {
    if (!pattern || !validationsText.length) {
      return false;
    }
    const conditions = pattern.map(rule => new RegExp(rule, 'g'));
    return !conditions.every(condition => condition.test(valueToValidate));
  };
  const handleOnChange = (newValue: string) => {
    setHasError(handleValidation(newValue));
    onChangeText(newValue);
  };

  return (
    <>
      {title ? (
        <CustomText
          text={title.toUpperCase()}
          color={isErrorEnabled ? COLORS.ALERT_DEFAULT : COLORS.GRAY_MEDIUM}
          bodyType="small-paragraph-bold"
        />
      ) : null}
      <View
        style={{
          ...styles.inputContainer,
          borderColor: isErrorEnabled
            ? COLORS.ALERT_DEFAULT
            : COLORS.GRAY_MEDIUM,
          ...(!!height && { height }),
        }}
      >
        <TextInput
          placeholder={placeholder}
          style={{
            ...styles.input,
            ...(!!height && { height }),
            color: isErrorEnabled ? COLORS.ALERT_DEFAULT : COLORS.BLACK,
          }}
          placeholderTextColor={
            isErrorEnabled ? COLORS.ALERT_DEFAULT : COLORS.GRAY_MEDIUM
          }
          autoCorrect={false}
          value={value}
          onChangeText={handleOnChange}
          multiline={textArea}
          numberOfLines={textArea ? 10 : 1}
          textAlignVertical={textArea ? 'top' : 'auto'}
          secureTextEntry={password}
          autoCapitalize="none"
        />
        {icon && <View style={styles.iconContainer}>{icon}</View>}
      </View>
      {isRequiredEnabled || isValidationEnabled ? (
        <View style={styles.errorContainer}>
          {isRequiredEnabled ? (
            <CustomText
              text={`- ${t('general:translation.requiredField')}`}
              color={COLORS.ALERT_DEFAULT}
              bodyType="small-paragraph"
            />
          ) : null}
          {isValidationEnabled
            ? validationsText.map(validation => (
                <CustomText
                  key={validation}
                  text={`- ${validation}`}
                  color={COLORS.ALERT_DEFAULT}
                  bodyType="small-paragraph"
                />
              ))
            : null}
        </View>
      ) : null}
    </>
  );
};

export default CustomInput;
