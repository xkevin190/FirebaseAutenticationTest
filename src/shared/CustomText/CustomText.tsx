import React from 'react';
import {Text, TextStyle} from 'react-native';

import {COLORS} from '../../constants/styles';
import {TextBody} from '../../types/CustomText';
import styles from './CustomText.style';

export interface ICustomTextProps {
  text: string;
  center?: boolean;
  color?: string;
  bodyType?: TextBody;
  testID?: string;
  customStyle?: TextStyle;
}

const CustomText: React.FC<ICustomTextProps> = ({
  text,
  color = COLORS.BLACK,
  bodyType = 'paragraph',
  center = false,
  testID = '',
  customStyle = {},
}) => (
  <Text
    style={{
      ...styles[bodyType],
      ...customStyle,
      ...(center && {textAlign: 'center'}),
      color,
    }}
    testID={testID}>
    {text}
  </Text>
);

export default CustomText;
