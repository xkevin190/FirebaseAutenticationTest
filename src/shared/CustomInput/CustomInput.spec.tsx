import React from 'react';
import {
  TextInputProps,
  TextStyle,
  ViewProps,
  ViewStyle,
  Text,
  TextProps,
} from 'react-native';
import { shallow, ShallowWrapper } from 'enzyme';

import { ICustomTextProps } from '../CustomText/CustomText';
import { COLORS } from '../../constants/styles';
import { emailRegex } from '../../utils/regex';
import CustomInput, { ICustomInputProps } from './CustomInput';

describe('CustomInput (Unit-Test)', () => {
  const props: ICustomInputProps = {
    value: 'TEST_INPUT',
    placeholder: 'TEST_PLACEHOLDER',
    onChangeText: jest.fn(),
  };
  let wrapper: ShallowWrapper;

  describe('general functionality', () => {
    beforeEach(() => {
      wrapper = shallow(<CustomInput {...props} />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders transparent background button to hide keyboard', () => {
      const button = wrapper.find('TouchableWithoutFeedback').at(0);
      expect(button).toBeDefined();
    });

    it('should render TextInput with "placeholder" prop', () => {
      const inputProps = wrapper
        .find('TextInput')
        .at(0)
        .props() as TextInputProps;
      expect(inputProps.placeholder).toEqual(props.placeholder);
    });

    it('should render TextInput with "GRAY_MEDIUM" placeholder color', () => {
      const inputProps = wrapper
        .find('TextInput')
        .at(0)
        .props() as TextInputProps;
      expect(inputProps.placeholderTextColor).toEqual(COLORS.GRAY_MEDIUM);
    });

    it('should render TextInput with "BLACK" color', () => {
      const inputProps = wrapper
        .find('TextInput')
        .at(0)
        .props() as TextInputProps;
      const inputStyles = inputProps.style as TextStyle;
      expect(inputStyles.color).toEqual(COLORS.BLACK);
    });

    it('should render TextInput with "GRAY_MEDIUM" border color', () => {
      const containerProps = wrapper.find('View').at(0).props() as ViewProps;
      const containerStyles = containerProps.style as ViewStyle;
      expect(containerStyles.borderColor).toEqual(COLORS.GRAY_MEDIUM);
    });

    it('should render TextInput with "value" prop', () => {
      const inputProps = wrapper
        .find('TextInput')
        .at(0)
        .props() as TextInputProps;
      expect(inputProps.value).toEqual(props.value);
    });

    it('should render TextInput with "multiline" false', () => {
      const inputProps = wrapper
        .find('TextInput')
        .at(0)
        .props() as TextInputProps;
      expect(inputProps.multiline).toBeFalsy();
    });

    it('should call "onChangeText" prop once with new "value" when changing text', () => {
      const input = wrapper.find('TextInput').at(0);
      const newValue = 'TEST_INPUT_CHANGED';
      input.simulate('changeText', newValue);
      expect(props.onChangeText).toBeCalledTimes(1);
      expect(props.onChangeText).toBeCalledWith(newValue);
    });

    it('should render TextInput with "numberOfLines" as one', () => {
      const inputProps = wrapper
        .find('TextInput')
        .at(0)
        .props() as TextInputProps;
      expect(inputProps.numberOfLines).toEqual(1);
    });

    it('should render TextInput with "textAlignVertical" as "auto"', () => {
      const inputProps = wrapper
        .find('TextInput')
        .at(0)
        .props() as TextInputProps;
      expect(inputProps.textAlignVertical).toEqual('auto');
    });
  });

  describe('when optional props are received', () => {
    it('should render title when "title" prop is received', () => {
      const title = 'TEST_TITLE_INPUT';
      wrapper = shallow(<CustomInput {...props} title={title} />);
      const text = wrapper.find('CustomText').at(0).props() as ICustomTextProps;
      expect(text.text).toEqual(title);
      expect(text.color).toEqual(COLORS.GRAY_MEDIUM);
    });

    it('should render TextInput with alert color and required field when "required" prop is received', () => {
      const title = 'TEST_TITLE_INPUT';
      wrapper = shallow(
        <CustomInput {...props} title={title} required value="" />,
      );
      const titleTextProps = wrapper
        .find('CustomText')
        .at(0)
        .props() as ICustomTextProps;
      const requiredTextProps = wrapper
        .find('CustomText')
        .at(1)
        .props() as ICustomTextProps;
      const inputProps = wrapper
        .find('TextInput')
        .at(0)
        .props() as TextInputProps;
      const containerProps = wrapper.find('View').at(0).props() as ViewProps;
      const inputStyles = inputProps.style as TextStyle;
      const containerStyles = containerProps.style as ViewStyle;
      expect(titleTextProps.text).toEqual(title);
      expect(titleTextProps.color).toEqual(COLORS.ALERT_DEFAULT);
      expect(inputStyles.color).toEqual(COLORS.ALERT_DEFAULT);
      expect(containerStyles.borderColor).toEqual(COLORS.ALERT_DEFAULT);
      expect(inputProps.placeholderTextColor).toEqual(COLORS.ALERT_DEFAULT);
      expect(requiredTextProps.color).toEqual(COLORS.ALERT_DEFAULT);
      expect(requiredTextProps.text).toEqual(
        '- general:translation.requiredField',
      );
    });

    it('should render TextInput with "height" prop', () => {
      const height = 10;
      wrapper = shallow(<CustomInput {...props} height={height} />);
      const inputProps = wrapper
        .find('TextInput')
        .at(0)
        .props() as TextInputProps;
      const inputStyles = inputProps.style as TextStyle;
      expect(inputStyles.height).toEqual(height);
    });

    it('should render TextInput with "multiline", "numberOfLines" as ten and textAlignVertical as "top" when "textArea" prop is recieved', () => {
      wrapper = shallow(<CustomInput {...props} textArea />);
      const inputProps = wrapper
        .find('TextInput')
        .at(0)
        .props() as TextInputProps;
      expect(inputProps.multiline).toBeTruthy();
      expect(inputProps.numberOfLines).toEqual(10);
      expect(inputProps.textAlignVertical).toEqual('top');
    });

    it('should render validation field when "pattern" and "validationsText" prop are received', () => {
      const validationsText = ['TEST_VALIDATION'];
      wrapper = shallow(
        <CustomInput
          {...props}
          pattern={[emailRegex]}
          validationsText={validationsText}
        />,
      );
      const input = wrapper.find('TextInput').at(0);
      const newValue = 'TEST_EMAIL_INCORRECT';
      input.simulate('changeText', newValue);
      const textProps = wrapper
        .find('CustomText')
        .at(0)
        .props() as ICustomTextProps;
      expect(textProps.text).toEqual('- TEST_VALIDATION');
      expect(textProps.color).toEqual(COLORS.ALERT_DEFAULT);
    });

    it('should render icon when "icon" prop recieved', () => {
      wrapper = shallow(
        <CustomInput {...props} icon={<Text>TEST_ICON</Text>} />,
      );
      const iconProps = wrapper.find('Text').at(0).props() as TextProps;
      expect(iconProps.children).toEqual('TEST_ICON');
    });
  });
});
