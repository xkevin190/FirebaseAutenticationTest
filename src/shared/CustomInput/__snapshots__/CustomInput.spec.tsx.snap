// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomInput (Unit-Test) general functionality should match snapshot 1`] = `
<Fragment>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "borderColor": "#9DA5B9",
        "borderRadius": 16,
        "borderWidth": 1,
        "flexDirection": "row",
        "marginBottom": 18,
        "marginTop": 6,
        "width": "100%",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      autoCorrect={false}
      multiline={false}
      numberOfLines={1}
      onChangeText={[Function]}
      placeholder="TEST_PLACEHOLDER"
      placeholderTextColor="#9DA5B9"
      secureTextEntry={false}
      style={
        Object {
          "color": "#000000",
          "flex": 1,
          "fontFamily": "Lato-Regular",
          "fontSize": 14,
          "paddingHorizontal": 16,
          "paddingVertical": 10,
        }
      }
      textAlignVertical="auto"
      value="TEST_INPUT"
    />
  </View>
</Fragment>
`;
